README.md

```diff
@@ -92,6 +92,66 @@ The callback function is passed a results object, containing the following keys:
 #### importer (starting from v2)
 `importer` is a `Function` to be called when libsass parser encounters the import directive. If present, libsass will call node-sass and let the user change file, data or both during the compilation. This option is optional, and applies to both render and renderSync functions. Also, it can either return object of form `{file:'..', contents: '..'}` or send it back via `done({})`. Note in renderSync or render, there is no restriction imposed on using `done()` callback or `return` statement (dispite of the asnchrony difference).
 
+#### functions
+`functions` is an `Object` that holds a collection of custom functions that may be invoked by the sass files being compiled. They may take zero or more input parameters and must return a value either synchronously (`return ...;`) or asynchronously (`done();`). Those parameters will be instances of one of the constructors contained in the `require('node-sass').types` hash. The return value must be of one of these types as well. See the list of available types below:
+
+##### types.Number(value [, unit = ""])
+* `getValue()`/ `setValue(value)` : gets / sets the numerical portion of the number
+* `getUnit()` / `setUnit(unit)` : gets / sets the unit portion of the number
+
+##### types.String(value)
+* `getValue()` / `setValue(value)` : gets / sets the enclosed string
+
+##### types.Color(r, g, b [, a = 1.0]) or types.Color(argb)
+* `getR()` / `setR(value)` : red component (integer from `0` to `255`)
+* `getG()` / `setG(value)` : green component (integer from `0` to `255`)
+* `getB()` / `setB(value)` : blue component (integer from `0` to `255`)
+* `getA()` / `setA(value)` : alpha component (number from `0` to `1.0`)
+
+Example:
+
+```javascript
+var Color = require('node-sass').types.Color,
+	c1 = new Color(255, 0, 0),
+	c2 = new Color(0xff0088cc);
+```
+
+##### types.Boolean(value)
+* `getValue()` / `setValue(value)` : gets / sets the enclosed boolean
+
+##### types.List(length [, commaSeparator = true])
+* `getValue(index)` / `setValue(index, value)` : `value` must itself be an instance of one of the constructors in `sass.types`.
+* `getSeparator()` / `setSeparator(isComma)` : whether to use commas as a separator
+* `getLength()`
+
+##### types.Map(length)
+* `getKey(index)` / `setKey(index, value)`
+* `getValue(index)` / `setValue(index, value)`
+* `getLength()`
+
+##### types.Null()
+
+##### Example
+
+```javascript
+sass.renderSync({
+  data: '#{headings(2,5)} { color: #08c; }',
+  functions: {
+    'headings($from: 0, $to: 6)': function(from, to) {
+      var i, f = from.getValue(), t = to.getValue(),
+          list = new sass.types.List(t - f + 1);
+
+      for (i = f; i <= t; i++) {
+        list.setValue(i - f, new sass.types.String('h' + i));
+      }
+
+      return list;
+    }
+  }
+});
+```
```

Comment:
@chriseppstein This is great input! You are right, the `SassTypes::Value` objects exposed to the v8 runtime constitute a very thin binding layer to the corresponding `Sass_Value` objects from `libsass` and as such expose all the same setters.

That said, custom functions are invoked with their arguments passed by value (i.e. `SassTypes::Value`’s constructor performs a deep copy of the underlying `Sass_Value` object via `sass_clone_value`) and similarly the results are returned by value (see `SassTypes::Value::get_sass_value()`). As a consequence it would not be possible for a custom function to mutate sass values created from a sass file. This immutable virtue is thus preserved for all values from the perspective of the sass files.

I understand that Sass as a language only exposes immutable values and as such it would make a lot of sense to hold onto this philosophy when it comes to custom functions especially if you guys think this would relieve ambiguities potentially perceived by the end developer.

However, If I can offer my two cents as a developer who spends a good deal of time working with JS: this is by no means a necessity. The fact that Sass’ design does not favor mutable values does not have to bleed onto the conventions established in Javascript. While I can totally make sense of immutable strings and numbers, I believe immutable maps and lists would feel awkward to a lot of JS developers out there. Additionally making these complex types immutable may lead to inferior performance results (and I said “may” because I have not looked at what’s behind the public libsass API, so I couldn't affirm that for sure).

In any case, I believe you are absolutely right in pushing for more user-friendly objects. And while it would make sense to me to leave all the subclasses of `SassTypes::Value` strictly as a translation layer, I can totally see a variety of facade objects in the Javascript realm that would expose all the goodies that you mentioned.

As far as the `Map` type is concerned, you would iterate over the map the same way you would if you had to use `libsass` directly:

```javascript
function(myMap) {
	var i, l, key, value;

	for (i = 0, l = myMap.getLength(); i < l; i++) {
		key = myMap.getKey(i);
		value = myMap.getValue(i);
	}
}
```